content/docs/sync-async-boundary.md

```diff
@@ -0,0 +1,43 @@
+
+## Sync
+
+The most popular medium for building synchronous systems is HTTP. At the topmost level the user's action is transformed into a HTTP request by the code running in the browser and submitted to one of the backend components for processing. The response carries the result of the action for rendering on screen. Internally the component that does the processing can delegate part of the work to other components. It does so by sending HTTP its own requests and collecting responses. These other components can do the same. This can result in the request "tree" that can span dozens of components.
+
+## Async
+
+In an asynchronous system components communicate by sending one-way messages to each other via message queues. The user's action is captured in a message that is put onto a queue. The message is picked up by another component. Processing a message can result in sending out more messages. Finally, the user interface component can receive a message that captures the result of the whole multi-step operation.
+
+## Trade-offs
+
+The synchronous approach allows a more straightforward programming model as the HTTP-based interactions can be represented as function calls returning values (or raising errors). It can be usually assumed that synchronous calls return quick and immediate responses. This allows building simple user interfaces that block until response comes back.
+
+The main problem with the synchronous approach is the temporal coupling it introduces. In order to process the request all the components involved need to be available at the same time. If processing the initial request results in more request (which in turn generate more requests) it can quickly get out of hand. The more parties need to be available at the same time, the higher the likelihood of any one of them failing and undermining the whole process.
+
+The asynchronous approach solves this problem by introducing queues and one-way messages. The sender or publisher puts the message onto a queue or topic and can immediately move on to processing subseqent work items. It does not waste time and resources actively waiting for a response. If the recipient is unavailable that OK too. The queue will ensure the message is not lost and that it will be eventually delivered. 
+
+This huge advantage of asynchronous systems becomes a disadvantage when it comes to building user interface components. The time to process a message is generally longer and more varied than in a synchronous system. In order for the user to know the state of the system, it might be necessary to visualize pending operations related to messages that have been sent but not yet processed. This introduces the need to keep the state on the client side. Depending on the complexity of the problem, it can lead to the need to implement fully-fledged multi-master replication as multiple clients can send messages that compete to modify the same set of data.
```

Comment:
```suggestion
This huge advantage of asynchronous systems becomes a disadvantage when it comes to building user interface components. The time to process a message is generally longer and more involved than in a synchronous system. In order for the user to know the state of the system, it might be necessary to visualize pending operations related to messages that have been sent but not yet processed. This introduces the need to keep the state client-side. Depending on the complexity of the problem, it can lead to a fully-fledged multi-master replication as multiple clients can send messages that compete to modify the same set of data.